File: deno/packages/training/cloud/together.ts
/**
 * Together AI fine-tuning provider.
 *
 * Equivalent to Rust's `brainwires_training::cloud::together` module.
 * Implements the same FineTuneProvider contract as OpenAI over Together's
 * `/v1/files` + `/v1/fine-tunes` endpoints.
 */

import { TrainingError } from "../error.ts";
import {
  DatasetId,
  type TrainingJobId,
  TrainingJobId as TrainingJobIdClass,
  type TrainingJobStatus,
  type TrainingJobSummary,
} from "../types.ts";
import type { CloudFineTuneConfig, DataFormat, FineTuneProvider } from "./types.ts";

export const TOGETHER_API_BASE = "https://api.together.xyz/v1";

export class TogetherFineTune implements FineTuneProvider {
  readonly name = "together";
  readonly base_url: string;
  private readonly api_key: string;

  constructor(api_key: string, base_url: string = TOGETHER_API_BASE) {
    this.api_key = api_key;
    this.base_url = base_url;
  }

  supportedBaseModels(): string[] {
    return [
      "meta-llama/Llama-3-8b-chat-hf",
      "meta-llama/Llama-3-70b-chat-hf",
      "mistralai/Mistral-7B-Instruct-v0.2",
      "Qwen/Qwen2-7B-Instruct",
    ];
  }

  supportsDpo(): boolean {
    return true;
  }

  /** Map Together's status strings onto the neutral enum. Exposed for tests. */
  static parseJobStatus(
    status_str: string,
    body: Record<string, unknown>,
  ): TrainingJobStatus {
    switch (status_str) {
      case "pending":
        return { status: "pending" };
      case "queued":
        return { status: "queued" };
      case "validating":
        return { status: "validating" };
      case "training":
      case "running":
        return {
          status: "running",
          progress: {
            epoch: 0,
            total_epochs: 0,
            step: typeof body.step === "number" ? body.step : 0,
            total_steps: typeof body.total_steps === "number" ? body.total_steps : 0,
            train_loss: null,
            eval_loss: null,
            learning_rate: null,
            elapsed_secs: 0,
          },
        };
      case "completed":
      case "succeeded": {
        const model_id = typeof body.output_name === "string" ? body.output_name : "unknown";
        return { status: "succeeded", model_id };
      }
      case "failed": {
        const err = body.error;
        const msg = typeof err === "string" ? err : (err as { message?: string } | undefined)?.message;
        return { status: "failed", error: msg ?? "Unknown error" };
      }
      case "cancelled":
      case "user_cancelled":
        return { status: "cancelled" };
      default:
        return { status: "pending" };
    }
  }

  async uploadDataset(data: Uint8Array, _format: DataFormat): Promise<DatasetId> {
    const form = new FormData();
    form.append("purpose", "fine-tune");
    form.append(
      "file",
      new Blob([data as BlobPart], { type: "application/json" }),
      "training_data.jsonl",
    );
    const res = await fetch(`${this.base_url}/files`, {
      method: "POST",
      headers: { Authorization: `Bearer ${this.api_key}` },
      body: form,
    });
    const body = await res.json() as Record<string, unknown>;
    if (!res.ok) throw TrainingError.api(errorMessage(body, "Together upload error"), res.status);
    const id = typeof body.id === "string" ? body.id : null;
    if (id === null) throw TrainingError.upload("Missing file ID in response");
    return new DatasetId(id);
  }

  async createJob(config: CloudFineTuneConfig): Promise<TrainingJobId> {
    const body: Record<string, unknown> = {
      training_file: config.training_dataset.value,
      model: config.base_model,
      n_epochs: config.hyperparams.epochs,
      batch_size: config.hyperparams.batch_size,
      learning_rate: config.hyperparams.learning_rate,
    };
    if (config.validation_dataset) body.validation_file = config.validation_dataset.value;
    if (config.suffix) body.suffix = config.suffix;
    if (config.lora) {
      body.lora = true;
      body.lora_r = config.lora.rank;
      body.lora_alpha = config.lora.alpha;
      body.lora_dropout = config.lora.dropout;
    }

    const res = await fetch(`${this.base_url}/fine-tunes`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.api_key}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    const resBody = await res.json() as Record<string, unknown>;
    if (!res.ok) throw TrainingError.api(errorMessage(resBody, "Unknown error"), res.status);
    const id = typeof resBody.id === "string" ? resBody.id : null;
    if (id === null) throw TrainingError.provider("Missing job ID in response");
    return new TrainingJobIdClass(id);
  }

  async getJobStatus(job_id: TrainingJobId): Promise<TrainingJobStatus> {
    const res = await fetch(`${this.base_url}/fine-tunes/${job_id.value}`, {
      headers: { Authorization: `Bearer ${this.api_key}` },
    });
    const body = await res.json() as Record<string, unknown>;
    if (!res.ok) {
      if (res.status === 404) throw TrainingError.jobNotFound(job_id.value);
      throw TrainingError.api(errorMessage(body, "Unknown error"), res.status);
    }
    const status_str = typeof body.status === "string" ? body.status : "unknown";
    return TogetherFineTune.parseJobStatus(status_str, body);
  }

  async cancelJob(job_id: TrainingJobId): Promise<void> {
    const res = await fetch(`${this.base_url}/fine-tunes/${job_id.value}/cancel`, {
      method: "POST",
      headers: { Authorization: `Bearer ${this.api_key}` },
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({})) as Record<string, unknown>;
      throw TrainingError.api(errorMessage(body, "Failed to cancel job"), res.status);
    }
  }

  async listJobs(): Promise<TrainingJobSummary[]> {
    const res = await fetch(`${this.base_url}/fine-tunes`, {
      headers: { Authorization: `Bearer ${this.api_key}` },
    });
    const body = await res.json() as Record<string, unknown>;
    if (!res.ok) throw TrainingError.api(errorMessage(body, "Failed to list jobs"), res.status);
    const data = Array.isArray(body.data) ? body.data : [];
    const out: TrainingJobSummary[] = [];
    for (const raw of data) {
      const job = raw as Record<string, unknown>;
      const id = job.id;
      const model = job.model;
      const status_str = job.status;
      if (typeof id !== "string" || typeof model !== "string" || typeof status_str !== "string") continue;
      const created = typeof job.created_at === "number"
        ? new Date(job.created_at * 1000).toISOString()
        : new Date().toISOString();
      out.push({
        job_id: new TrainingJobIdClass(id),
        provider: "together",
        base_model: model,
        status: TogetherFineTune.parseJobStatus(status_str, job),
        created_at: created,
        metrics: null,
      });
    }
    return out;
  }

  async deleteModel(model_id: string): Promise<void> {
    const res = await fetch(`${this.base_url}/models/${model_id}`, {
      method: "DELETE",
      headers: { Authorization: `Bearer ${this.api_key}` },
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({})) as Record<string, unknown>;
      throw TrainingError.api(errorMessage(body, "Failed to delete model"), res.status);
    }
  }
}

function errorMessage(body: Record<string, unknown>, fallback: string): string {
  const err = body.error;
  if (typeof err === "string") return err;
  if (err && typeof err === "object") {
    const m = (err as Record<string, unknown>).message;
    if (typeof m === "string") return m;
  }
  const msg = body.message;
  return typeof msg === "string" ? msg : fallback;
}

File: crates/brainwires-finetune/src/cloud/together.rs
use async_trait::async_trait;
use reqwest::Client;
use serde_json::json;
use tracing::debug;

use crate::datasets::DataFormat;

use super::{CloudFineTuneConfig, FineTuneProvider};
use crate::error::TrainingError;
use crate::types::{
    DatasetId, TrainingJobId, TrainingJobStatus, TrainingJobSummary, TrainingProgress,
};

const TOGETHER_API_URL: &str = "https://api.together.xyz/v1";

/// Together AI fine-tuning provider.
///
/// Supports SFT and DPO fine-tuning with OpenAI-compatible format.
pub struct TogetherFineTune {
    api_key: String,
    client: Client,
    base_url: String,
}

impl TogetherFineTune {
    /// Create a new Together AI fine-tune provider.
    pub fn new(api_key: impl Into<String>) -> Self {
        Self {
            api_key: api_key.into(),
            client: Client::new(),
            base_url: TOGETHER_API_URL.to_string(),
        }
    }

    /// Set a custom base URL.
    pub fn with_base_url(mut self, url: impl Into<String>) -> Self {
        self.base_url = url.into();
        self
    }

    /// Extract error message from API response body.
    fn extract_error(body: &serde_json::Value) -> String {
        body.get("error")
            .and_then(|e| {
                // Try {"error": {"message": "..."}} first, then {"error": "..."}
                e.get("message")
                    .and_then(|m| m.as_str())
                    .or_else(|| e.as_str())
            })
            .unwrap_or("Unknown error")
            .to_string()
    }

    /// Parse job status from API response.
    fn parse_job_status(body: &serde_json::Value) -> TrainingJobStatus {
        let status_str = body
            .get("status")
            .and_then(|v| v.as_str())
            .unwrap_or("pending");

        match status_str {
            "pending" | "queued" => TrainingJobStatus::Queued,
            "running" | "processing" => TrainingJobStatus::Running {
                progress: TrainingProgress::default(),
            },
            "completed" | "succeeded" => {
                let model_id = body
                    .get("output_name")
                    .or_else(|| body.get("fine_tuned_model"))
                    .and_then(|v| v.as_str())
                    .unwrap_or("unknown")
                    .to_string();
                TrainingJobStatus::Succeeded { model_id }
            }
            "failed" | "error" => TrainingJobStatus::Failed {
                error: Self::extract_error(body),
            },
            "cancelled" => TrainingJobStatus::Cancelled,
            _ => TrainingJobStatus::Pending,
        }
    }
}

#[async_trait]
impl FineTuneProvider for TogetherFineTune {
    fn name(&self) -> &str {
        "together"
    }

    fn supported_base_models(&self) -> Vec<String> {
        vec![
            "meta-llama/Meta-Llama-3.1-8B-Instruct".to_string(),
            "meta-llama/Meta-Llama-3.1-70B-Instruct".to_string(),
            "mistralai/Mixtral-8x7B-Instruct-v0.1".to_string(),
            "mistralai/Mistral-7B-Instruct-v0.3".to_string(),
            "Qwen/Qwen2.5-7B-Instruct".to_string(),
        ]
    }

    fn supports_dpo(&self) -> bool {
        true
    }

    async fn upload_dataset(
        &self,
        data: &[u8],
        _format: DataFormat,
    ) -> Result<DatasetId, TrainingError> {
        debug!("Uploading dataset to Together AI ({} bytes)", data.len());

        let part = reqwest::multipart::Part::bytes(data.to_vec()).file_name("training_data.jsonl");

        let form = reqwest::multipart::Form::new()
            .text("purpose", "fine-tune")
            .part("file", part);

        let response = self
            .client
            .post(format!("{}/files", self.base_url))
            .bearer_auth(&self.api_key)
            .multipart(form)
            .send()
            .await?;

        let status = response.status();
        let body: serde_json::Value = response.json().await?;

        if !status.is_success() {
            return Err(TrainingError::Api {
                message: Self::extract_error(&body),
                status_code: status.as_u16(),
            });
        }

        let file_id = body
            .get("id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| TrainingError::Upload("Missing file ID".to_string()))?
            .to_string();

        Ok(DatasetId(file_id))
    }

    async fn create_job(
        &self,
        config: CloudFineTuneConfig,
    ) -> Result<TrainingJobId, TrainingError> {
        debug!(
            "Creating Together AI fine-tuning job for: {}",
            config.base_model
        );

        let mut body = json!({
            "training_file": config.training_dataset.0,
            "model": config.base_model,
            "n_epochs": config.hyperparams.epochs,
            "learning_rate": config.hyperparams.learning_rate,
            "batch_size": config.hyperparams.batch_size,
        });

        if let Some(ref suffix) = config.suffix {
            body["suffix"] = json!(suffix);
        }

        let response = self
            .client
            .post(format!("{}/fine-tunes", self.base_url))
            .bearer_auth(&self.api_key)
            .json(&body)
            .send()
            .await?;

        let status = response.status();
        let response_body: serde_json::Value = response.json().await?;

        if !status.is_success() {
            return Err(TrainingError::Api {
                message: Self::extract_error(&response_body),
                status_code: status.as_u16(),
            });
        }

        let job_id = response_body
            .get("id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| TrainingError::Provider("Missing job ID".to_string()))?
            .to_string();

        Ok(TrainingJobId(job_id))
    }

    async fn get_job_status(
        &self,
        job_id: &TrainingJobId,
    ) -> Result<TrainingJobStatus, TrainingError> {
        let url = format!("{}/fine-tunes/{}", self.base_url, job_id.0);

        let response = self
            .client
            .get(&url)
            .bearer_auth(&self.api_key)
            .send()
            .await?;

        let status = response.status();
        let body: serde_json::Value = response.json().await?;

        if !status.is_success() {
            return Err(TrainingError::Api {
                message: Self::extract_error(&body),
                status_code: status.as_u16(),
            });
        }

        Ok(Self::parse_job_status(&body))
    }

    async fn cancel_job(&self, job_id: &TrainingJobId) -> Result<(), TrainingError> {
        let url = format!("{}/fine-tunes/{}/cancel", self.base_url, job_id.0);

        let response = self
            .client
            .post(&url)
            .bearer_auth(&self.api_key)
            .send()
            .await?;

        if !response.status().is_success() {
            let body: serde_json::Value = response.json().await.unwrap_or_default();
            return Err(TrainingError::Provider(Self::extract_error(&body)));
        }

        Ok(())
    }

    async fn list_jobs(&self) -> Result<Vec<TrainingJobSummary>, TrainingError> {
        let response = self
            .client
            .get(format!("{}/fine-tunes", self.base_url))
            .bearer_auth(&self.api_key)
            .send()
            .await?;

        let body: serde_json::Value = response.json().await?;
        let data = body.get("data").and_then(|v| v.as_array());

        Ok(data
            .map(|jobs| {
                jobs.iter()
                    .filter_map(|j| {
                        Some(TrainingJobSummary {
                            job_id: TrainingJobId(j.get("id")?.as_str()?.to_string()),
                            provider: "together".to_string(),
                            base_model: j.get("model")?.as_str()?.to_string(),
                            status: Self::parse_job_status(j),
                            created_at: chrono::Utc::now(),
                            metrics: None,
                        })
                    })
                    .collect()
            })
            .unwrap_or_default())
    }

    async fn delete_model(&self, model_id: &str) -> Result<(), TrainingError> {
        let url = format!("{}/models/{}", self.base_url, model_id);

        let response = self
            .client
            .delete(&url)
            .bearer_auth(&self.api_key)
            .send()
            .await?;

        if !response.status().is_success() {
            let body: serde_json::Value = response.json().await.unwrap_or_default();
            return Err(TrainingError::Provider(Self::extract_error(&body)));
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_together_supported_models() {
        let provider = TogetherFineTune::new("test");
        assert!(provider.supports_dpo());
        assert!(!provider.supported_base_models().is_empty());
    }
}
